### ISSUE TITLE
Question of speed and some methods

### ISSUE THREAD

**Issue by SZ-Chander** [ml-explore/mlx-examples#17]
Thanks for this work, I've been looking forward to the framework running on Apple silicon for a long time!
The device I'm using is a MacBook Pro-14ich by M3Max.
I was surprised to find that the mnist project using the mlx framework(main.py) runs slower than the Pytorch(torch_main.py). Can I ask what causes this? Pytorch is a great framework and I use it all the time for learning, but it still surprises me that mlx as a dedicated framework would be slower than Pytorch.
As well, does the mlx framework currently have or have plans to develop some tools like torch.utils.data.Dataset, torch.utils.data.DataLoader classes and the torchvision? I just tried to reproduce a yolo algorithm using mlx, but I didn't find these methods in the documentation.
Finally, looking forward to the continued development of mlx, I'm very much looking forward to the future that UMA and dedicated frameworks can create!

**Comment by AjaniStewart** [comment]
Somewhat related, the MNIST example seems to run faster on cpu than on gpu. This is on a 13in M1 MacBook Pro.

```
(mlx) ➜  mnist git:(main) python main.py --gpu
Epoch 0: Test accuracy 0.866, Time 0.302 (s)
Epoch 1: Test accuracy 0.896, Time 0.295 (s)
Epoch 2: Test accuracy 0.917, Time 0.304 (s)
Epoch 3: Test accuracy 0.928, Time 0.307 (s)
Epoch 4: Test accuracy 0.935, Time 0.311 (s)
Epoch 5: Test accuracy 0.938, Time 0.341 (s)
Epoch 6: Test accuracy 0.946, Time 0.348 (s)
Epoch 7: Test accuracy 0.941, Time 0.354 (s)
Epoch 8: Test accuracy 0.950, Time 0.368 (s)
Epoch 9: Test accuracy 0.948, Time 0.367 (s)
(mlx) ➜  mnist git:(main) python main.py      
Epoch 0: Test accuracy 0.866, Time 0.098 (s)
Epoch 1: Test accuracy 0.896, Time 0.094 (s)
Epoch 2: Test accuracy 0.917, Time 0.100 (s)
Epoch 3: Test accuracy 0.928, Time 0.096 (s)
Epoch 4: Test accuracy 0.935, Time 0.094 (s)
Epoch 5: Test accuracy 0.938, Time 0.095 (s)
Epoch 6: Test accuracy 0.946, Time 0.098 (s)
Epoch 7: Test accuracy 0.941, Time 0.093 (s)
Epoch 8: Test accuracy 0.950, Time 0.094 (s)
Epoch 9: Test accuracy 0.948, Time 0.094 (s)
```

**Comment by radarFudan** [comment]
Similar problem for 16-inch M1Pro with 32GB memory.

Using 13.6.2 (22G320)

(mlx) ➜  mnist git:(main) python3 main.py      
Epoch 0: Test accuracy 0.866, Time 0.107 (s)
Epoch 1: Test accuracy 0.896, Time 0.106 (s)
Epoch 2: Test accuracy 0.917, Time 0.117 (s)
Epoch 3: Test accuracy 0.928, Time 0.116 (s)
Epoch 4: Test accuracy 0.935, Time 0.114 (s)
Epoch 5: Test accuracy 0.938, Time 0.114 (s)
Epoch 6: Test accuracy 0.946, Time 0.112 (s)
Epoch 7: Test accuracy 0.941, Time 0.112 (s)
Epoch 8: Test accuracy 0.950, Time 0.113 (s)
Epoch 9: Test accuracy 0.948, Time 0.105 (s)
(mlx) ➜  mnist git:(main) python3 main.py --gpu
Epoch 0: Test accuracy 0.866, Time 0.227 (s)
Epoch 1: Test accuracy 0.896, Time 0.207 (s)
Epoch 2: Test accuracy 0.917, Time 0.202 (s)
Epoch 3: Test accuracy 0.928, Time 0.223 (s)
Epoch 4: Test accuracy 0.935, Time 0.198 (s)
Epoch 5: Test accuracy 0.938, Time 0.199 (s)
Epoch 6: Test accuracy 0.946, Time 0.199 (s)
Epoch 7: Test accuracy 0.941, Time 0.202 (s)
Epoch 8: Test accuracy 0.950, Time 0.208 (s)
Epoch 9: Test accuracy 0.948, Time 0.198 (s)

**Comment by Weizhongjin** [comment]
similar problem in new 14 cpu core 30 gpu core m3max macbookpro, 
python main.py
Epoch 0: Test accuracy 0.866, Time 0.211 (s)
Epoch 1: Test accuracy 0.896, Time 0.080 (s)
Epoch 2: Test accuracy 0.917, Time 0.083 (s)
Epoch 3: Test accuracy 0.928, Time 0.082 (s)
Epoch 4: Test accuracy 0.935, Time 0.079 (s)
Epoch 5: Test accuracy 0.938, Time 0.080 (s)
Epoch 6: Test accuracy 0.946, Time 0.082 (s)
Epoch 7: Test accuracy 0.941, Time 0.082 (s)
Epoch 8: Test accuracy 0.950, Time 0.082 (s)
Epoch 9: Test accuracy 0.948, Time 0.082 (s)

-------------------------------------------------
python main.py --gpu
Epoch 0: Test accuracy 0.866, Time 0.280 (s)
Epoch 1: Test accuracy 0.896, Time 0.133 (s)
Epoch 2: Test accuracy 0.917, Time 0.143 (s)
Epoch 3: Test accuracy 0.928, Time 0.132 (s)
Epoch 4: Test accuracy 0.935, Time 0.134 (s)
Epoch 5: Test accuracy 0.938, Time 0.135 (s)
Epoch 6: Test accuracy 0.946, Time 0.138 (s)
Epoch 7: Test accuracy 0.941, Time 0.123 (s)
Epoch 8: Test accuracy 0.950, Time 0.130 (s)
Epoch 9: Test accuracy 0.948, Time 0.132 (s)

**Comment by awni** [comment]
I don't know if I would call that a problem per se. The CPU + coprocessors are very very fast, and they have less overhead than running ops on the GPU. So for such compute-light workloads the CPU actually makes a lot of sense as the default processor.

You should try our transformers or LLMs examples on  CPU vs GPU to see how the GPU can be much faster.

Regarding being slower than PyTorch make sure you are comparing apples to apples (e.g. both running on CPU or GPU):

```
python main.py --gpu
python torch_main.py --gpu
```

On my machine the second is more than twice slower than the first. On the CPU the examples are closer but MLX is still slightly faster.

Closing this issue for now. Feel free to comment if you are observing any other strange behavior and I will reopen.
